Error while loading transcripts (build 37)

```
273247', 'GRCh37'), ('4', 140334956, 140337300, 'ENST00000608345', 'ENSG00000273247', 'GRCh37'), ('4', 140334970, 140374930, 'ENST00000608663', 'ENSG00000273247', 'GRCh37'), ('4', 140334978, 140375188, 'ENST00000609359', 'ENSG00000273247', 'GRCh37')  ... displaying 10 of 13 total bound parameter sets ...  ('1', 148864594, 148865175, 'ENST00000435754', 'ENSG00000242943', 'GRCh37'), ('6', 102901242, 102901667, 'ENST00000407792', 'ENSG00000219757', 'GRCh37')]
2025-02-13 08:47:17,618 INFO sqlalchemy.engine.Engine COMMIT
[2025-02-13 08:47:17 +0100] [14] [ERROR] Exception in ASGI application
Traceback (most recent call last):
  File "/venv/lib/python3.11/site-packages/requests/models.py", line 820, in generate
    yield from self.raw.stream(chunk_size, decode_content=True)
  File "/venv/lib/python3.11/site-packages/urllib3/response.py", line 1063, in stream
    yield from self.read_chunked(amt, decode_content=decode_content)
  File "/venv/lib/python3.11/site-packages/urllib3/response.py", line 1219, in read_chunked
    self._update_chunk_length()
  File "/venv/lib/python3.11/site-packages/urllib3/response.py", line 1149, in _update_chunk_length
    raise ProtocolError("Response ended prematurely") from None
urllib3.exceptions.ProtocolError: Response ended prematurely

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/worker/app/src/chanjo2/endpoints/intervals.py", line 98, in load_transcripts
    nr_loaded_transcripts: int = await update_transcripts(
                                 ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/worker/app/src/chanjo2/meta/handle_load_intervals.py", line 156, in update_transcripts
    for line in lines:
  File "/venv/lib/python3.11/site-packages/requests/models.py", line 869, in iter_lines
    for chunk in self.iter_content(
  File "/venv/lib/python3.11/site-packages/requests/utils.py", line 572, in stream_decode_response_unicode
    for chunk in iterator:
  File "/venv/lib/python3.11/site-packages/requests/models.py", line 822, in generate
    raise ChunkedEncodingError(e)
requests.exceptions.ChunkedEncodingError: Response ended prematurely

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/venv/lib/python3.11/site-packages/starlette/_exception_handler.py", line 51, in wrapped_app
    await app(scope, receive, sender)
  File "/venv/lib/python3.11/site-packages/starlette/routing.py", line 73, in app
    response = await f(request)
               ^^^^^^^^^^^^^^^^
  File "/venv/lib/python3.11/site-packages/fastapi/routing.py", line 301, in app
    raw_response = await run_endpoint_function(
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/venv/lib/python3.11/site-packages/fastapi/routing.py", line 212, in run_endpoint_function
    return await dependant.call(**values)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/worker/app/src/chanjo2/endpoints/intervals.py", line 108, in load_transcripts
    raise HTTPException(
fastapi.exceptions.HTTPException: 422: (ProtocolError('Response ended prematurely'),)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/venv/lib/python3.11/site-packages/uvicorn/protocols/http/h11_impl.py", line 429, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/venv/lib/python3.11/site-packages/uvicorn/middleware/proxy_headers.py", line 78, in __call__
    return await self.app(scope, receive, send)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/venv/lib/python3.11/site-packages/fastapi/applications.py", line 1054, in __call__
    await super().__call__(scope, receive, send)
  File "/venv/lib/python3.11/site-packages/starlette/applications.py", line 113, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/venv/lib/python3.11/site-packages/starlette/middleware/errors.py", line 187, in __call__
    raise exc
  File "/venv/lib/python3.11/site-packages/starlette/middleware/errors.py", line 165, in __call__
    await self.app(scope, receive, _send)
  File "/venv/lib/python3.11/site-packages/starlette/middleware/exceptions.py", line 62, in __call__
    await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
  File "/venv/lib/python3.11/site-packages/starlette/_exception_handler.py", line 62, in wrapped_app
    raise exc
  File "/venv/lib/python3.11/site-packages/starlette/_exception_handler.py", line 51, in wrapped_app
    await app(scope, receive, sender)
  File "/venv/lib/python3.11/site-packages/starlette/routing.py", line 715, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/venv/lib/python3.11/site-packages/starlette/routing.py", line 735, in app
    await route.handle(scope, receive, send)
  File "/venv/lib/python3.11/site-packages/starlette/routing.py", line 288, in handle
    await self.app(scope, receive, send)
  File "/venv/lib/python3.11/site-packages/starlette/routing.py", line 76, in app
    await wrap_app_handling_exceptions(app, request)(scope, receive, send)
  File "/venv/lib/python3.11/site-packages/starlette/_exception_handler.py", line 73, in wrapped_app
    response = await handler(conn, exc)
               ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/venv/lib/python3.11/site-packages/fastapi/exception_handlers.py", line 15, in http_exception_handler
    return JSONResponse(
           ^^^^^^^^^^^^^
  File "/venv/lib/python3.11/site-packages/starlette/responses.py", line 180, in __init__
    super().__init__(content, status_code, headers, media_type, background)
  File "/venv/lib/python3.11/site-packages/starlette/responses.py", line 43, in __init__
    self.body = self.render(content)
                ^^^^^^^^^^^^^^^^^^^^
  File "/venv/lib/python3.11/site-packages/starlette/responses.py", line 183, in render
    return json.dumps(
           ^^^^^^^^^^^
  File "/usr/local/lib/python3.11/json/__init__.py", line 238, in dumps
    **kw).encode(obj)
          ^^^^^^^^^^^
  File "/usr/local/lib/python3.11/json/encoder.py", line 200, in encode
    chunks = self.iterencode(o, _one_shot=True)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/json/encoder.py", line 258, in iterencode
    return _iterencode(o, 0)
           ^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/json/encoder.py", line 180, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type ProtocolError is not JSON serializable
[hiseq.clinical@cg-services-prod ~]$
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,7 +2,7 @@
 ### Changed
 - Updated schug library to v1.10
 - Disable SQLAlchemy logger
-- Update genes, transcripts exons only from pre-downloaded files from schug
+- To avoid timeout errors, update genes, transcripts exons only from pre-downloaded files from schug
 
 ## [3.1]
 ### Changed

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,6 +2,7 @@
 ### Changed
 - Updated schug library to v1.10
 - Disable SQLAlchemy logger
+- Update genes, transcripts exons only from pre-downloaded files from schug
 
 ## [3.1]
 ### Changed

--- a/src/chanjo2/crud/intervals.py
+++ b/src/chanjo2/crud/intervals.py
@@ -41,22 +41,9 @@ def _filter_intervals_by_build(
     return intervals.filter(interval_type.build == build)
 
 
-def create_db_gene(gene: GeneBase) -> SQLGene:
-    """Create and return a SQL gene object."""
-    return SQLGene(
-        build=gene.build.value,
-        chromosome=gene.chromosome,
-        start=gene.start,
-        stop=gene.stop,
-        ensembl_id=gene.ensembl_id,
-        hgnc_symbol=gene.hgnc_symbol,
-        hgnc_id=gene.hgnc_id,
-    )
-
-
-def bulk_insert_genes(db: Session, genes: List[GeneBase]):
+def bulk_insert_genes(db: Session, genes: List[SQLGene]):
     """Bulk insert genes into the database."""
-    db.bulk_save_objects([create_db_gene(gene) for gene in genes])
+    db.bulk_save_objects(genes)
     db.commit()
 
 
@@ -96,29 +83,9 @@ def get_genes(
     return genes.all()
 
 
-def create_db_transcript(transcript: TranscriptBase) -> SQLTranscript:
-    """Create and return a SQL transcript object."""
-
-    return SQLTranscript(
-        chromosome=transcript.chromosome,
-        start=transcript.start,
-        stop=transcript.stop,
-        ensembl_id=transcript.ensembl_id,
-        ensembl_gene_id=transcript.ensembl_gene_id,
-        refseq_mrna=transcript.refseq_mrna,
-        refseq_mrna_pred=transcript.refseq_mrna_pred,
-        refseq_ncrna=transcript.refseq_ncrna,
-        refseq_mane_select=transcript.refseq_mane_select,
-        refseq_mane_plus_clinical=transcript.refseq_mane_plus_clinical,
-        build=transcript.build,
-    )
-
-
 def bulk_insert_transcripts(db: Session, transcripts: List[TranscriptBase]):
     """Bulk insert transcripts into the database."""
-    db.bulk_save_objects(
-        [create_db_transcript(transcript=transcript) for transcript in transcripts]
-    )
+    db.bulk_save_objects(transcripts)
     db.commit()
 
 
@@ -221,22 +188,7 @@ def get_gene_intervals(
     return intervals.all()
 
 
-def create_db_exon(exon: ExonBase) -> SQLExon:
-    """Create and return a SQL exon object."""
-
-    return SQLExon(
-        chromosome=exon.chromosome,
-        start=exon.start,
-        stop=exon.stop,
-        rank_in_transcript=exon.rank_in_transcript,
-        ensembl_id=exon.ensembl_id,
-        ensembl_gene_id=exon.ensembl_gene_id,
-        ensembl_transcript_id=exon.ensembl_transcript_id,
-        build=exon.build,
-    )
-
-
-def bulk_insert_exons(db: Session, exons: List[ExonBase]) -> None:
+def bulk_insert_exons(db: Session, exons: List[SQLExon]) -> None:
     """Bulk insert exons into the database."""
-    db.bulk_save_objects([create_db_exon(exon=exon) for exon in exons])
+    db.bulk_save_objects(exons)
     db.commit()

--- a/src/chanjo2/endpoints/intervals.py
+++ b/src/chanjo2/endpoints/intervals.py
@@ -34,19 +34,16 @@ def count_nr_filters(filters: List[str]) -> int:
 @router.post("/intervals/load/genes/{build}")
 async def load_genes(
     build: Builds,
-    file_path: Optional[str] = None,
+    file_path: str,
     session: Session = Depends(get_session),
 ) -> Response:
     """Load genes in the given genome build."""
 
     try:
-        if file_path:
-            gene_lines: Iterator[str] = resource_lines(file_path=file_path)
-            nr_loaded_genes: int = await update_genes(
-                build=build, lines=gene_lines, session=session
-            )
-        else:
-            nr_loaded_genes: int = await update_genes(build=build, session=session)
+        gene_lines: Iterator[str] = resource_lines(file_path=file_path)
+        nr_loaded_genes: int = await update_genes(
+            build=build, lines=gene_lines, session=session
+        )
         return JSONResponse(
             content={"detail": f"{nr_loaded_genes} genes loaded into the database"}
         )
@@ -83,21 +80,16 @@ async def genes(query: GeneQuery, session: Session = Depends(get_session)):
 @router.post("/intervals/load/transcripts/{build}", response_model=List[GeneBase])
 async def load_transcripts(
     build: Builds,
-    file_path: Optional[str] = None,
+    file_path: str,
     session: Session = Depends(get_session),
 ) -> Response:
     """Load transcripts in the given genome build."""
 
     try:
-        if file_path:
-            transcripts_lines: Iterator[str] = resource_lines(file_path=file_path)
-            nr_loaded_transcripts: int = await update_transcripts(
-                build=build, lines=transcripts_lines, session=session
-            )
-        else:
-            nr_loaded_transcripts: int = await update_transcripts(
-                build=build, session=session
-            )
+        transcripts_lines: Iterator[str] = resource_lines(file_path=file_path)
+        nr_loaded_transcripts: int = await update_transcripts(
+            build=build, lines=transcripts_lines, session=session
+        )
         return JSONResponse(
             content={
                 "detail": f"{nr_loaded_transcripts} transcripts loaded into the database"
@@ -144,19 +136,16 @@ async def transcripts(
 @router.post("/intervals/load/exons/{build}")
 async def load_exons(
     build: Builds,
-    file_path: Optional[str] = None,
+    file_path: str,
     session: Session = Depends(get_session),
 ) -> Response:
     """Load exons in the given genome build."""
 
     try:
-        if file_path:
-            exons_lines: Iterator[str] = resource_lines(file_path=file_path)
-            nr_loaded_exons: int = await update_exons(
-                build=build, lines=exons_lines, session=session
-            )
-        else:
-            nr_loaded_exons: int = await update_exons(build=build, session=session)
+        exons_lines: Iterator[str] = resource_lines(file_path=file_path)
+        nr_loaded_exons: int = await update_exons(
+            build=build, lines=exons_lines, session=session
+        )
         return JSONResponse(
             content={"detail": f"{nr_loaded_exons} exons loaded into the database"}
         )

--- a/src/chanjo2/meta/handle_load_intervals.py
+++ b/src/chanjo2/meta/handle_load_intervals.py
@@ -59,7 +59,7 @@ async def update_genes(
 ) -> Optional[int]:
     """Loads genes into the database."""
 
-    LOG.info(f"Loading gene intervals. Genome build --> {build}")
+    LOG.warning(f"Updating genes. Genome build --> {build}")
     if lines is None:
         lines: Iterator[str] = read_resource_lines(
             build=build, interval_type=IntervalType.GENES
@@ -72,6 +72,7 @@ async def update_genes(
             f"Ensembl genes file has an unexpected format:{header}. Expected format: {GENES_FILE_HEADER[build]}"
         )
 
+    LOG.warning(f"Deleting genes, transcripts, exons in build {build}")
     for interval_type in [SQLExon, SQLTranscript, SQLGene]:
         delete_intervals_for_build(db=session, interval_type=interval_type, build=build)
 
@@ -82,7 +83,8 @@ async def update_genes(
             break
 
         items: List = _replace_empty_cols(line=line, nr_expected_columns=len(header))
-
+        if items == header:
+            continue
         try:
             sql_gene = SQLGene(
                 build=build,
@@ -118,7 +120,7 @@ async def update_transcripts(
 ) -> Optional[int]:
     """Loads transcripts into the database."""
 
-    LOG.info(f"Loading transcript intervals. Genome build --> {build}")
+    LOG.warning(f"Updating transcripts. Genome build --> {build}")
 
     if lines is None:
         lines: Iterator[str] = read_resource_lines(
@@ -132,6 +134,7 @@ async def update_transcripts(
             f"Ensembl transcripts file has an unexpected format:{header}. Expected format: {TRANSCRIPTS_FILE_HEADER[build]}"
         )
 
+    LOG.warning(f"Deleting transcripts in build {build}")
     delete_intervals_for_build(db=session, interval_type=SQLTranscript, build=build)
 
     transcripts_bulk: List[TranscriptBase] = []
@@ -141,9 +144,10 @@ async def update_transcripts(
             break
 
         items: List = _replace_empty_cols(line=line, nr_expected_columns=len(header))
-
+        if items == header:
+            continue
         try:
-            transcript = TranscriptBase(
+            transcript = SQLTranscript(
                 chromosome=items[0],
                 ensembl_gene_id=items[1],
                 ensembl_id=items[2],
@@ -184,7 +188,7 @@ async def update_exons(
 ) -> Optional[int]:
     """Loads exons into the database."""
 
-    LOG.info(f"Loading exon intervals. Genome build --> {build}")
+    LOG.warning(f"Updating exons. Genome build --> {build}")
 
     if lines is None:
         lines: Iterator[str] = read_resource_lines(
@@ -198,6 +202,7 @@ async def update_exons(
             f"Ensembl exons file has an unexpected format:{header}. Expected format: {EXONS_FILE_HEADER[build]}"
         )
 
+    LOG.warning(f"Deleting exons in build {build}")
     delete_intervals_for_build(db=session, interval_type=SQLExon, build=build)
 
     exons_bulk: List[ExonBase] = []
@@ -207,10 +212,11 @@ async def update_exons(
             break
 
         items: List = _replace_empty_cols(line=line, nr_expected_columns=len(header))
-
+        if items == header:
+            continue
         try:
             # Load Exon interval into the database
-            exon = ExonBase(
+            exon = SQLExon(
                 chromosome=items[0],
                 ensembl_gene_id=items[1],
                 ensembl_transcript_id=items[2],

--- a/tests/src/chanjo2/endpoints/test_intervals.py
+++ b/tests/src/chanjo2/endpoints/test_intervals.py
@@ -1,9 +1,8 @@
-from typing import Dict, Iterator, List, Type
+from typing import Dict, List, Type
 
 import pytest
 from fastapi import status
 from fastapi.testclient import TestClient
-from pytest_mock.plugin import MockerFixture
 
 from chanjo2.constants import MULTIPLE_PARAMS_NOT_SUPPORTED_MSG
 from chanjo2.meta.handle_bed import resource_lines
@@ -19,45 +18,6 @@ MOCKED_FILE_PARSER = "chanjo2.meta.handle_load_intervals.read_resource_lines"
 
 @pytest.mark.parametrize("build, path", BUILD_GENES_RESOURCE)
 def test_load_genes(
-    build: str,
-    path: str,
-    client: TestClient,
-    endpoints: Type,
-    mocker: MockerFixture,
-):
-    """Test the endpoint that adds genes to the database in a given genome build."""
-
-    # GIVEN a patched response from Ensembl Biomart, via schug
-    gene_lines: Iterator = resource_lines(path)
-    mocker.patch(
-        MOCKED_FILE_PARSER,
-        return_value=gene_lines,
-    )
-
-    # GIVEN a number of genes contained in the demo file
-    nr_genes: int = len(list(resource_lines(path))) - 1
-
-    # WHEN sending a request to the load_genes with genome build
-    response: Response = client.post(f"{endpoints.LOAD_GENES}{build}")
-
-    # THEN it should return success
-    assert response.status_code == status.HTTP_200_OK
-
-    # THEN all the genes should be loaded
-    assert response.json()["detail"] == f"{nr_genes} genes loaded into the database"
-
-    # WHEN sending a request to the "genes" endpoint
-    response: Response = client.post(endpoints.GENES, json={"build": build})
-    assert response.status_code == status.HTTP_200_OK
-    result = response.json()
-    # THEN the expected number of genes should be returned
-    assert len(result) == nr_genes
-    # AND the gene should have the right format
-    assert GeneBase(**result[0])
-
-
-@pytest.mark.parametrize("build, path", BUILD_GENES_RESOURCE)
-def test_load_genes_from_file(
     build: str,
     path: str,
     client: TestClient,
@@ -183,37 +143,6 @@ def test_genes_by_hgnc_symbols(
 
 @pytest.mark.parametrize("build, path", BUILD_TRANSCRIPTS_RESOURCE)
 def test_load_transcripts(
-    build: str,
-    path: str,
-    client: TestClient,
-    endpoints: Type,
-    mocker: MockerFixture,
-):
-    """Test the endpoint that adds genes to the database in a given genome build."""
-
-    # GIVEN a patched response from Ensembl Biomart, via schug
-    transcript_lines: Iterator = resource_lines(path)
-    mocker.patch(
-        MOCKED_FILE_PARSER,
-        return_value=transcript_lines,
-    )
-
-    # GIVEN a number of transcripts contained in the demo file
-    nr_transcripts: int = len(list(resource_lines(path))) - 1
-
-    # WHEN sending a request to the load_genes with genome build
-    response: Response = client.post(f"{endpoints.LOAD_TRANSCRIPTS}{build}")
-    # THEN it should return success
-    assert response.status_code == status.HTTP_200_OK
-    # THEN all transcripts should be loaded
-    assert (
-        response.json()["detail"]
-        == f"{nr_transcripts} transcripts loaded into the database"
-    )
-
-
-@pytest.mark.parametrize("build, path", BUILD_TRANSCRIPTS_RESOURCE)
-def test_load_transcripts_from_file(
     build: str,
     path: str,
     client: TestClient,
@@ -375,36 +304,7 @@ def test_transcripts_by_hgnc_symbols(
 
 
 @pytest.mark.parametrize("build, path", BUILD_EXONS_RESOURCE)
-def test_load_exons(
-    build: str,
-    path: str,
-    client: TestClient,
-    endpoints: Type,
-    mocker: MockerFixture,
-):
-    """Test the endpoint that adds exons to the database in a given genome build."""
-
-    # GIVEN a patched response from Ensembl Biomart, via schug
-    exons_lines: Iterator = resource_lines(path)
-    mocker.patch(
-        MOCKED_FILE_PARSER,
-        return_value=exons_lines,
-    )
-
-    # GIVEN a number of exons contained in the demo file
-    nr_exons: int = len(list(resource_lines(path))) - 1
-
-    # WHEN sending a request to the load_genes with genome build
-    response: Response = client.post(f"{endpoints.LOAD_EXONS}{build}")
-
-    # THEN it should return success
-    assert response.status_code == status.HTTP_200_OK
-    # THEN all exons should be loaded
-    assert response.json()["detail"] == f"{nr_exons} exons loaded into the database"
-
-
-@pytest.mark.parametrize("build, path", BUILD_EXONS_RESOURCE)
-def test_load_exons_from_file(
+def test_load_exons_from(
     build: str,
     path: str,
     client: TestClient,

